### PR TITLE
feat: image resizing

### PR DESCRIPTION
**Pull request by colincclark**
https://nidigitalsolutions.jira.com/browse/REPLAT-1711

- fixes general image resizing and moves it to the utils package (now a dependency)
- brings `utils` package coverage to 100%
- resizes images correctly within the image package for optimum effect
- more work for lazy loading and SSR to be done on https://github.com/newsuk/times-components/issues/871

Paired with @tuncaulubilge

### CODE REVIEW

File: packages/image/src/image.js [hunk 1]
```diff
@@ -9,8 +13,10 @@ class TimesImage extends Component {
   constructor(props) {
     super(props);
 
+    console.log(screenWidthInPixels(), normaliseWidth(screenWidthInPixels()));
```
Comment (craigbilner) [inline packages/image/src/image.js]: should have failed linting?

File: packages/author-profile/__tests__/web/author-profile-content.web.test.js [hunk 2]
```diff
@@ -151,7 +151,7 @@ describe("AuthorProfile test on web", () => {
         .at(0)
         .props().uri
     ).toEqual(
-      "//www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2F1b5afe88-cb0d-11e7-9ee9-e45ae7e1cdd4.jpg?crop=4252%2C2835%2C0%2C0&resize=100"
+      "//www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2F1b5afe88-cb0d-11e7-9ee9-e45ae7e1cdd4.jpg?crop=4252%2C2835%2C0%2C0"
```
Comment (craigbilner) [inline packages/author-profile/__tests__/web/author-profile-content.web.test.js]: this looks to have broken

File: packages/author-profile/__tests__/ios/__snapshots__/author-profile.test.js.snap [hunk 3]
```diff
@@ -2361,7 +2361,7 @@ exports[`AuthorProfile test on ios renders profile content component 1`] = `
                       onLoad={[Function]}
                       source={
                         Object {
-                          "uri": "https://www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2F1b5afe88-cb0d-11e7-9ee9-e45ae7e1cdd4.jpg?crop=4252%2C2835%2C0%2C0&resize=660",
+                          "uri": "https://www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2F1b5afe88-cb0d-11e7-9ee9-e45ae7e1cdd4.jpg?crop=4252%2C2835%2C0%2C0&resize=1440&resize=1440",
```
Comment (craigbilner) [inline packages/author-profile/__tests__/ios/__snapshots__/author-profile.test.js.snap]: double trouble :100:

File: packages/image/src/image.js (hunk 1 above)
Comment (colincclark) [reply packages/image/src/image.js]: removed

File: packages/author-profile/__tests__/web/author-profile-content.web.test.js (hunk 2 above)
Comment (colincclark) [reply packages/author-profile/__tests__/web/author-profile-content.web.test.js]: fixed

File: packages/article/__tests__/android/__snapshots__/article.android.test.js.snap [hunk 4]
```diff
@@ -1261,7 +1261,7 @@ exports[`Article tests on android should render a full article 1`] = `
                           onLoad={[Function]}
                           source={
                             Object {
-                              "uri": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F5f243dd6-93aa-11e7-a2ce-ce94682a575d.jpg?crop=3000%2C1687%2C0%2C156&resize=1200&preview=true",
+                              "uri": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F5f243dd6-93aa-11e7-a2ce-ce94682a575d.jpg?crop=3000%2C1687%2C0%2C156&preview=true",
```
Comment (tuncaulubilge) [inline packages/article/__tests__/android/__snapshots__/article.android.test.js.snap]: This should have been removed, as I deleted image.android.js that creates preview urls

File: packages/article/__tests__/android/__snapshots__/article.android.test.js.snap [hunk 5]
```diff
@@ -5772,8 +5772,8 @@ exports[`Article tests on android should render an article with a video asset 1`
             accessibilityLabel="splash-component"
             style={
               Object {
-                "height": 421.875,
-                "width": 750,
+                "height": NaN,
+                "width": undefined,
```
Comment (tuncaulubilge) [inline packages/article/__tests__/android/__snapshots__/article.android.test.js.snap]: Something wrong here, will take a look

File: packages/article/__tests__/ios/__snapshots__/article.ios.test.js.snap [hunk 6]
```diff
@@ -39,7 +39,7 @@ exports[`Article tests on ios should render a full article 1`] = `
           "credits": null,
           "crop": Object {
             "ratio": "16:9",
-            "url": "//www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F5f243dd6-93aa-11e7-a2ce-ce94682a575d.jpg?crop=3000%2C1687%2C0%2C156&resize=1200",
+            "url": "//www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F5f243dd6-93aa-11e7-a2ce-ce94682a575d.jpg?crop=3000%2C1687%2C0%2C156",
```
Comment (tuncaulubilge) [inline packages/article/__tests__/ios/__snapshots__/article.ios.test.js.snap]: We shouldn't have any native images without resize, will take a look at this

File: packages/article/__tests__/android/__snapshots__/article.android.test.js.snap (hunk 4 above)
Comment (colincclark) [reply packages/article/__tests__/android/__snapshots__/article.android.test.js.snap]: have removed from dist file and fixed snapshots

File: packages/author-profile/__tests__/ios/__snapshots__/author-profile.test.js.snap (hunk 3 above)
Comment (colincclark) [reply packages/author-profile/__tests__/ios/__snapshots__/author-profile.test.js.snap]: @craigbilner this is caused by `resize` being added to image in card. is there still a reason to do that?

Comment (craigbilner) [reply packages/author-profile/__tests__/ios/__snapshots__/author-profile.test.js.snap]: nope, can move wherever it makes sense

Comment (tuncaulubilge) [reply packages/author-profile/__tests__/ios/__snapshots__/author-profile.test.js.snap]: Fixed by 9bde775

File: packages/article/__tests__/android/__snapshots__/article.android.test.js.snap (hunk 5 above)
Comment (tuncaulubilge) [reply packages/article/__tests__/android/__snapshots__/article.android.test.js.snap]: Fixed by 0da3c7c

**Comment by colincclark** [comment]
@craigbilner please re-review. We have a bunch of tickets to write from this spike / refactor

**Comment by colincclark** [comment]
@craigbilner can you unblock? we will do the work for https://github.com/newsuk/times-components/issues/871 once it has been planned in

File: packages/author-profile/src/author-profile-content.web.js [hunk 7]
```diff
@@ -225,7 +225,7 @@ class AuthorProfileContent extends Component {
                           <AuthorProfileItem
                             {...article}
                             imageRatio={imageRatio}
-                            imageSize={this.getImageSize(elementId)}
+                            imageSize={this.getImageSize(elementId) || 100}
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/author-profile/src/author-profile-content.web.js]: Moved it here from the default imageSize value of Card component. The purpose is to render a low-res version of an image and lazy-load the actual image. Colin created a github issue so lazy-loading logic will be moved into image.

File: packages/image/src/image.js [hunk 8]
```diff
@@ -30,8 +35,10 @@ class TimesImage extends Component {
       onLoad: this.handleLoad
     };
 
-    if (uri) {
-      props.source = { uri };
+    if (uri && width > 0) {
+      props.source = {
+        uri: `${uri}&resize=${width}`
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/image/src/image.js]: All image urls has a crop value as a query param already, so this shouldn't be a problem

File: packages/image/src/image.js [hunk 9]
```diff
@@ -10,7 +14,8 @@ class TimesImage extends Component {
     super(props);
 
     this.state = {
-      isLoaded: false
+      isLoaded: false,
+      width: normaliseWidth(screenWidthInPixels())
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/image/src/image.js]: Configuration changes such as orientation change could change this value, so preferred to use state even if we don't support orientation change at the moment. I can gladly move it to render if you think that would be better

[the remark this answers is not in the document]
Comment (colincclark) [reply packages/image/src/image.js]: Let's leave the orientation change support in, previous commit history indicates this was intended to be supported

File: packages/ad/package.json [hunk 10]
```diff
@@ -69,6 +68,7 @@
   },
   "dependencies": {
     "@times-components/styleguide": "0.5.16",
+    "@times-components/utils": "0.6.13",
```
Comment (craigbilner) [inline packages/ad/package.json]: can we remove all ad changes?

File: packages/article/__tests__/android/__snapshots__/article.android.test.js.snap [hunk 11]
```diff
@@ -39,7 +39,7 @@ exports[`Article tests on android should render a full article 1`] = `
           "credits": null,
           "crop": Object {
             "ratio": "16:9",
-            "url": "//www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F5f243dd6-93aa-11e7-a2ce-ce94682a575d.jpg?crop=3000%2C1687%2C0%2C156&resize=1200",
+            "url": "//www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F5f243dd6-93aa-11e7-a2ce-ce94682a575d.jpg?crop=3000%2C1687%2C0%2C156",
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/article/__tests__/android/__snapshots__/article.android.test.js.snap]: This is just the raw data inside the snapshot. We removed resize parameters from the fixtures, that caused this change

File: packages/article/__tests__/android/__snapshots__/article.android.test.js.snap [hunk 12]
```diff
@@ -2431,7 +2365,7 @@ exports[`Article tests on android should render a full article 1`] = `
                               onLoad={[Function]}
                               source={
                                 Object {
-                                  "uri": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fb7ac990a-d872-11e7-9b16-edd3826708d6.jpg?crop=3959%2C2639%2C0%2C0&preview=true",
+                                  "uri": "https://www.thetimes.co.uk/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fb7ac990a-d872-11e7-9b16-edd3826708d6.jpg?crop=3959%2C2639%2C0%2C0&resize=1440",
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/article/__tests__/android/__snapshots__/article.android.test.js.snap]: Jest seems to be mocking Dimensions and PixelRatio, which results of 1500 pixels for the screen width. In reality, only the high end devices should use 1440px for images

File: packages/author-head/__tests__/web/__snapshots__/author-head.web.test.js.snap [hunk 13]
```diff
@@ -202,7 +202,7 @@ exports[`AuthorHead test on web renders with data 1`] = `
           >
             <img
               alt=""
-              src="https://feeds.thetimes.co.uk/web/imageserver/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F0694e84e-04ff-11e7-976a-0b4b9a1a67a3.jpg?crop=854,854,214,0&resize=200"
+              src="https://feeds.thetimes.co.uk/web/imageserver/imageserver/image/methode%2Ftimes%2Fprod%2Fweb%2Fbin%2F0694e84e-04ff-11e7-976a-0b4b9a1a67a3.jpg?crop=854,854,214,0"
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/author-head/__tests__/web/__snapshots__/author-head.web.test.js.snap]: Probably some old fixtures, will take a look at this :+1:

File: packages/related-articles/__tests__/ios/__snapshots__/related-articles.test.js.snap [hunk 14]
```diff
@@ -141,7 +141,7 @@ exports[`Related articles tests on ios:  Lead and two template should render one
                       onLoad={[Function]}
                       source={
                         Object {
-                          "uri": "https://www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fe42e32fe-d14c-11e7-b1ec-8503a5941b97.jpg?crop=6250%2C3516%2C0%2C326&resize=996",
+                          "uri": "https://www.thetimes.co.uk/imageserver/image/%2Fmethode%2Ftimes%2Fprod%2Fweb%2Fbin%2Fe42e32fe-d14c-11e7-b1ec-8503a5941b97.jpg?crop=6250%2C3516%2C0%2C326&resize=996&resize=1440",
```
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/related-articles/__tests__/ios/__snapshots__/related-articles.test.js.snap]: Thought we got rid of the magic 996, will double check :+1:

File: packages/author-head/__tests__/web/__snapshots__/author-head.web.test.js.snap (hunk 13 above)
[the remark this answers is not in the document]
Comment (tuncaulubilge) [reply packages/author-head/__tests__/web/__snapshots__/author-head.web.test.js.snap]: Seems to be some old fixture data in author-head

**Pull request merged**